### ISSUE TITLE
Ceph "Cluster utilization" chart no longer working on dashboard after v18.2.0 upgrade

### ISSUE THREAD

**Issue by sethjones** [rook/rook#12876]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
When navigating to the homepage/dashboard in the Ceph Dashboard UI, the "Cluster utilization" chart is not longer functional.  Charts consistently act as if they have no data.

**Expected behavior:**
Dashboard chart loads and displays current values based on platform metrics.

**How to reproduce it (minimal and precise):**

While on the latest version of the Rook operator (deployed via helm), upgrade Ceph from v17 --> v18.2.0.  No additional settings changes made, except image update.

**File(s) to submit**:
Dashboard screenshot:
![Screenshot 2023-09-09 at 09-24-03 Ceph](https://github.com/rook/rook/assets/2685052/2bab6c51-59b2-4f27-a7dd-63c7b7276ecd)

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
```yaml
spec:
  cephVersion:
    allowUnsupported: false
    image: quay.io/ceph/ceph:v18.2.0
  cleanupPolicy:
    allowUninstallWithVolumes: false
    confirmation: ''
    sanitizeDisks:
      dataSource: zero
      iteration: 1
      method: quick
  continueUpgradeAfterChecksEvenIfNotHealthy: false
  crashCollector:
    disable: false
  dashboard:
    enabled: true
    port: 8443
    ssl: false
  dataDirHostPath: /var/lib/rook
  disruptionManagement:
    managePodBudgets: true
    osdMaintenanceTimeout: 30
    pgHealthCheckTimeout: 0
  external: {}
  healthCheck:
    daemonHealth:
      mon:
        disabled: false
        interval: 45s
      osd:
        disabled: false
        interval: 60s
      status:
        disabled: false
        interval: 60s
    livenessProbe:
      mgr:
        disabled: false
      mon:
        disabled: false
      osd:
        disabled: false
    startupProbe:
      mgr:
        disabled: false
      mon:
        disabled: false
      osd:
        disabled: false
  logCollector:
    enabled: true
    maxLogSize: 500M
    periodicity: daily
  mgr:
    allowMultiplePerNode: false
    count: 2
    modules:
      - enabled: true
        name: pg_autoscaler
  mon:
    allowMultiplePerNode: false
    count: 3
  monitoring:
    enabled: false
  network:
    connections:
      compression:
        enabled: false
      encryption:
        enabled: false
  priorityClassNames:
    mgr: system-cluster-critical
    mon: system-node-critical
    osd: system-node-critical
  removeOSDsIfOutAndSafeToRemove: false
  security:
    kms: {}
  skipUpgradeChecks: false
  storage:
    onlyApplyOSDPlacement: false
    useAllDevices: true
    useAllNodes: true
  waitTimeoutForHealthyOSDInMinutes: 10

```

**Logs to submit**:
No known evidence of error in operator or manager logs.

**Cluster Status to submit**:
n/a

**Environment**:
* OS (e.g. from /etc/os-release): Ubuntu 22.04
* Kernel (e.g. `uname -a`): Linux rke2-ceph-0 5.15.0-1041-kvm #46-Ubuntu SMP Fri Aug 25 07:39:11 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
* Cloud provider or hardware configuration: Proxmox VMs
* Rook version (use `rook version` inside of a Rook Pod): v1.12.3
* Storage backend version (e.g. for ceph do `ceph -v`): v18.2.0
* Kubernetes version (use `kubectl version`):  v1.26.7
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): rke2
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)):

**Comment by travisn** [comment]
@avanthakkar @jmolmo @rkachach Is this a known issue with the Reef dashboard? Several Rook users have reported this issue. Who can look into it? thanks

**Comment by nizamial09** [comment]
The `Cluster Utiliziation` data's depends on the prometheus. So these values will only show up when there is prometheus configureed. @sethjones Is the prometheus configured in the cluster?

**Comment by sethjones** [comment]
> The `Cluster Utiliziation` data's depends on the prometheus. So these values will only show up when there is prometheus configureed. @sethjones Is the prometheus configured in the cluster?

Yes, I am using the default RKE2 prometheus instance.  This dashboard was functioning normally prior to the v18.2.0 upgrade, without changes to the cluster or monitoring setup.

**Comment by nizamial09** [comment]
Okay, then probably this should be a bug in the dashboard. We'll need to test this with the rook orch to see what went wrong there. While we do that, you can still use the old dashboard as your default. You can follow this doc which explains [how to switch to your old dashboard](https://docs.ceph.com/en/latest/mgr/dashboard/#overview-of-the-dashboard-landing-page), there is a `Note` section which says that. Or you can simply issue `ceph dashboard feature disable dashboard` from the toolbox to do that as well. 

The landing page is still new and we are still improving it. Hopefully we'll fix all these bugs and release a stable version of it soon.

**Comment by nizamial09** [comment]
meanwhile i opened a tracker for the team to track it on ceph: https://tracker.ceph.com/issues/62821

**Comment by nizamial09** [comment]
I managed to test this locally with a rook cluster today and the reason chart is not populating is because the dashboard is trying to consume the prometheus metrics but it doesn't know where the prometheus metrics are. 

Dashboard has a setting called `PROMETHEUS_API_HOST` which needs to be updated with the url of the prometheus endpoint. In my case it was
```
╰─$ kc get -n rook-ceph services rook-prometheus
NAME              TYPE       CLUSTER-IP      EXTERNAL-IP   PORT(S)          AGE
rook-prometheus   NodePort   10.108.222.61   <none>        9090:30900/TCP   15m
```
so, `http://10.108.222.61:9090`. You can update it like `ceph dashboard set-prometheus-api-host http://10.108.222.61:9090`. Then I refreshed the UI, wrote some IOs and saw the charts value.

![Screenshot from 2023-09-14 11-35-33](https://github.com/rook/rook/assets/71764184/da8d599b-8e20-4583-8cd9-d12af82f24fa)

@travisn, Is there a way rook can update the dashboard config options when deploying the prometheus service?

**Comment by nizamial09** [comment]
saw this: https://github.com/rook/rook/issues/11369

**Comment by electrical** [comment]
In our case we run a separate prometheus instance. do we need to do anything special for rook/ceph to push the metrics or is it automatically processed?

**Comment by travisn** [comment]
@nizamial09 Yes, Rook can set that value for the dashboard. It would be ideal if the operator could automatically detect this from a service with well-known prometheus labels. If that's not possible, I'm thinking we would need a setting in the CephCluster CR. Even if we could detect the endpoint automatically in some scenarios, this setting could allow overriding the endpoint in case auto detection is not working correctly. For example, the setting could be `dashboard.prometheusEndpoint`. This setting should also be mentioned in `Documentation/ceph-monitoring.md` so users enabling prometheus can find it. 

@rkachach Could you look into this?

**Comment by nizamial09** [comment]
> do we need to do anything special for rook/ceph to push the metrics or is it automatically processed?

@electrical: In rook, there are some service-monitors that you'll need to create which watches the rook cluster.
https://rook.io/docs/rook/latest/Storage-Configuration/Monitoring/ceph-monitoring/#prometheus-instances

You can check the doc to get more info.

**Comment by nizamial09** [comment]
> we would need a setting in the CephCluster CR. Even if we could detect the endpoint automatically in some scenarios, this setting could allow overriding the endpoint in case auto detection is not working correctly.

Yup, I would also prefer to have a setting for `prometheusEndpoint` which would give user the flexibility of editing the URL in case of unexpected issues. They can still do it from toolbox but I guess the rook users might prefer to do it from the crds as stated in https://github.com/rook/rook/issues/11369.

**Comment by electrical** [comment]
> > do we need to do anything special for rook/ceph to push the metrics or is it automatically processed?
> 
> @electrical: In rook, there are some service-monitors that you'll need to create which watches the cluster. https://rook.io/docs/rook/latest/Storage-Configuration/Monitoring/ceph-monitoring/#prometheus-instances
> 
> You can check the doc to get more info.

Thank you @nizamial09 seems we were already collecting the metrics but had to configure the dashboard to connect to the prometheus api.
Thank you!

**Comment by nizamial09** [comment]
@electrical: good to know that! you're welcome!

**Comment by rkachach** [comment]
> @nizamial09 Yes, Rook can set that value for the dashboard. It would be ideal if the operator could automatically detect this from a service with well-known prometheus labels. If that's not possible, I'm thinking we would need a setting in the CephCluster CR. Even if we could detect the endpoint automatically in some scenarios, this setting could allow overriding the endpoint in case auto detection is not working correctly. For example, the setting could be `dashboard.prometheusEndpoint`. This setting should also be mentioned in `Documentation/ceph-monitoring.md` so users enabling prometheus can find it.
> 
> @rkachach Could you look into this?

Sure, I'll talk to Nizam and see what changes will be required :+1:

**Comment by blackliner** [comment]
To configure your prometheus endpoint:
```
    dashboard:
      enabled: true
      port: 8080
      ssl: false
      prometheusEndpoint: http://kube-prometheus-stack-prometheus.kube-prometheus-stack.svc:9090
```

@nizamial09 Is this assumption correct: The old dashboards metrics did not require prometheus, but the new one has a dependency and out of the box it will not work? Shouldn't there at least be a fallback that uses the same data source as the old dashboard did?

**Comment by voarsh2** [comment]
> Okay, then probably this should be a bug in the dashboard. We'll need to test this with the rook orch to see what went wrong there. While we do that, you can still use the old dashboard as your default. You can follow this doc which explains [how to switch to your old dashboard](https://docs.ceph.com/en/latest/mgr/dashboard/#overview-of-the-dashboard-landing-page), there is a `Note` section which says that. Or you can simply issue `ceph dashboard feature disable dashboard` from the toolbox to do that as well.
> 
> The landing page is still new and we are still improving it. Hopefully we'll fix all these bugs and release a stable version of it soon.

I'm confused.
I get these basic stats with the old dashboard, without prometheus, but when I use the new dashboard, I get nothing. And then I do setup prometheus the dashboard doesn't seem to "use" it......

![image](https://github.com/rook/rook/assets/60437695/18032bfe-708c-4197-aa9f-264325b6b720)

**Comment by SanjayM0han** [comment]
Hello Ceph Community,

I recently upgraded my Ceph cluster through the following versions:

17.2.7 >> 18.2.0 >> 18.2.2 >> 18.2.4 >> 19.2.0 (Squid)
After successfully upgrading to Ceph 19.2.0, I noticed that the recovery throughput is no longer visible in the new Ceph dashboard. However, old dashboard features and metrics appear to be functioning correctly, and the recovery throughput was displayed properly in the previous version of the Ceph dashboard.

Old Dashboard

![96edbbc2-b053-4fb8-b9d5-f21a10007489](https://github.com/user-attachments/assets/88bd7d36-6773-4d1b-8612-3f1d884d6d6f)

New Dashboard

![aa9d73a0-df6d-4de9-963c-3880b99c39d7](https://github.com/user-attachments/assets/abe10c77-8355-47e2-8aff-263369b997f7)

Steps to reproduce:

Upgrade Ceph cluster from 17.2.7 to 19.2.0.
Access the Ceph dashboard.
Check for recovery throughput metrics in the new dashboard.
Current behavior:

Recovery throughput is not displayed in the new Ceph dashboard after upgrading.
Expected behavior:

Recovery throughput should be visible as it was in the previous versions of the Ceph dashboard.
Has anyone else encountered this issue, or are there any known fixes or workarounds? I would appreciate any insights or suggestions.
